Infinispan upstream community operator

galderz opened the pull request:
* Some metadata still to be added by @oraNod (see TBD).

I've managed to pushing this operator bundle using operator-courier to my [quay.io repository](https://quay.io/application/gzamarre/infinispan?tab=releases). 

I've then used an OpenShift 4 cluster I already have running with OLM (I'm aware that I need to test this on vanilla Kube but don't have such set up yet), to install [this OperatorSource](https://gist.github.com/galderz/a77a3e2434c097f5720d7eff8d1b472c#file-gzamarre-operatorsource-cr-yaml), but when I go to the OperatorHub UI I don't see my operator there. The [logs](https://gist.github.com/galderz/a77a3e2434c097f5720d7eff8d1b472c#file-oc-logs-opsrc) for the OperatorSource look ok, and I can see the operator in the list of my operator source. Am I missing something?

Also, once you have your operator source installed, how do you run the operator? I've tried to do:

```
$ oc apply -f https://raw.githubusercontent.com/infinispan/infinispan-operator/0.1.0/deploy/cr/cr_minimal.yaml
```

But got:

```
error: unable to recognize "https://raw.githubusercontent.com/infinispan/infinispan-operator/0.1.0/deploy/cr/cr_minimal.yaml": no matches for kind "Infinispan" in version "infinispan.org/v1"
```

galderz commented:
Btw, I can't find any info on how to test all of this with vanilla Kube, so if you could share that too that'd be great.

galderz commented:
Hmmm, actually, just spotted this:

```
$ oc get events
4m36s       Warning   Unhealthy   pod/gzamarre-operators-584df6f946-hlbck   Liveness probe errored: rpc error: code = Unknown desc = command error: command timed out, stdout: , stderr: , exit code -1
69m         Warning   Unhealthy   pod/gzamarre-operators-584df6f946-hlbck   Liveness probe failed: timeout: failed to connect service "localhost:50051" within 1s
```

galderz commented:
Looking at my operator pod log, the last line it shows is:

```
time="2019-03-14T09:39:53Z" level=info msg="loading Packages" port=50051 type=appregistry
```

Others show:

```
time="2019-03-13T08:52:15Z" level=info msg="loading Packages" port=50051 type=appregistry
time="2019-03-13T08:52:15Z" level=info msg="extracting provided API information" port=50051 type=appregistry
time="2019-03-13T08:52:15Z" level=info msg="serving registry" port=50051 type=appregistry
```

Looking into the package yaml file

galderz commented:
I've pushed a fix to the package yaml. I've pushed a new release to the quay.io and recreated the operator source but no changes, still see the unhealthy warnings and logs halting in same place.

SamiSousa commented:
Hi, @galderz! Thank you for submitting a PR. To answer your question about testing on vanilla kube, you can follow a similar flow to how you're testing with a custom OperatorSource on OpenShift. Once you've set up a kubernetes cluster with OLM installed, install the marketplace operator and install your operator by following the steps in the readme here: https://github.com/operator-framework/operator-marketplace#deploying-the-marketplace-operator-with-kubernetes

galderz commented:
@SamiSousa Ok. What about the issue I'm having? Seems like registry logging stops without any errors?

SamiSousa commented:
@galderz The errors you're seeing in your registry pod lead me to believe there may be issues in your manifests. Let me review this PR and see

SamiSousa commented on upstream-community-operators/infinispan/infinispan-operator.v0.1.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,181 @@
+            "size": 3
+          }
+        },
+        {
```
Not sure how OLM handles multiple examples for the same CR. @alecmerdler do you have some insight into this?

alecmerdler replied:
Console will just grab the first one for now.

galderz replied:
@SamiSousa etcd 0.9.2 operator shows multiple examples, I got the idea from there.

SamiSousa commented on upstream-community-operators/infinispan/infinispan-operator.v0.1.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,181 @@
+          }
+        }
+      ]
+    description: TBD
```
Leaving a comment here since if this is merged, this would be displayed in the operator tile in operatorhub.io

galderz replied:
This is not final submission, @oraNod and I will complete all `TBD` sections.

SamiSousa commented on upstream-community-operators/infinispan/infinispan-operator.v0.1.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,181 @@
+spec:
+  displayName: infinispan
+  description: |
+    TBD
```
Update description

SamiSousa commented on upstream-community-operators/infinispan/infinispan-operator.v0.1.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,181 @@
+  keywords: ['infinispan', 'key value', 'database', 'datagrid', 'open source']
+  version: 0.1.0
+  maturity: stable
+  replaces: infinispan-operator.v0.0.0
```
Please remove this. There is no CSV that matches this name in this PR, so this would cause OLM to fail installing this operator.

SamiSousa commented on upstream-community-operators/infinispan/infinispan-operator.v0.1.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,181 @@
+    - name: Infinispan Community
+      email: infinispan-dev@lists.jboss.org
+  provider:
+    name: Red Hat
```
Should this be `Infinispan` to match the maintainers and support?

SamiSousa commented on upstream-community-operators/infinispan/infinispan-operator.v0.1.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,181 @@
+      url: https://github.com/infinispan/infinispan-operator
+
+  icon:
+    - base64data: TBD
```
Please omit the icon if none is present. This might break the UI

galderz replied:
Commented for the time being. I don't want us to forget about it.

SamiSousa commented on upstream-community-operators/infinispan/infinispan-operator.v0.1.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,181 @@
+                      - name: WATCH_NAMESPACE
+                        valueFrom:
+                          fieldRef:
+                            fieldPath: metadata.namespace
```
If this operator supports the downward API for watching cluster resources, then this should be set to `metadata.annotations['olm.targetNamespaces']`

galderz replied:
We don't yet support watching the cluster resources, what should we put here instead?

galderz replied:
@SamiSousa I copied `metadata.namespace` from the etc operator. What's the difference between the two values? What is this really for?

SamiSousa replied:
If this operator doesn't support watching resources clusterwide, please set the `AllNamespaces` installMode to `false`.
This field should be set to `metadata.annotations['olm.targetNamespaces']`. The etcd operator is an antipattern.
Operators should be using OLM's annotation to know which namespace(s) to watch. If the operator is installed with an OperatorGroup that watches one namespace, this annotation will be set to the name of that one namespace. If the OperatorGroup watches multiple namespaces, this annotation will be set to a comma separated list of namespaces that are watched. If the OperatorGroup watches all namespaces, this string is set to the empty string, and the operator must use the downward API to watch CRs clusterwide.
More info can be found here: https://github.com/operator-framework/operator-lifecycle-manager/blob/master/Documentation/design/operatorgroups.md

galderz commented:
> Has this been tested on vanilla kubernetes?

Not yet. Since I had a running OpenShift 4 instance already with OLM, decided to try there first. Once that's working I'd try vanilla Kubernetes which requires more set up on my side.

galderz commented:
@SamiSousa Applied the suggested fixes (see last commit) and that made the operator pod start correctly. I can now see our operator on operator hub 🎉

galderz commented:
I'll sync up with @oraNod WRT the `TBD` sections and I'll test it in vanilla kubernetes next.

galderz commented:
Once you have your operator in the operator hub UI, how do you install it via command line? I've seen instructions [here](https://github.com/operator-framework/operator-marketplace#installing-an-operator-using-marketplace) but when I try to apply the catalog source, the pod keeps getting created/terminated with errors like:

```
3m55s       Warning   FailedCreatePodSandBox   pod/installed-infinispan-operators-7dc64cc985-xkklb    Failed create pod sandbox: rpc error: code = Unknown desc = failed to create pod network sandbox k8s_installed-infinispan-operators-7dc64cc985-xkklb_openshift-marketplace_15d63d65-499a-11e9-a142-025036b333c2_0(53be4ddb19ce304152523ca5d62151bdad272e2df90668c37fb595f062e37ed5): Multus: Err adding pod to network "openshift-sdn": cannot set "openshift-sdn" ifname to "eth0": no netns: failed to Statfs "/proc/62425/ns/net": no such file or directory
```

galderz commented:
Hmmm, seems like catalogsource is already created when the operatorsource was created? Do you really need it? In any case, I went and installed the operator via the UI and although it says it succeed, it's not among the list of installed operators. Any more details on how to go from having an operator in the operator hub UI to installing it and instantiating it?

galderz commented:
Eventually figured there's something called an installplan (not mentioned at all in testing-operators.md) which does not have a running pod. Instead, the install plan itself contains information about any issues, in this case:

```
        error creating csv infinispan-operator.v0.1.0: ClusterServiceVersion.operators.coreos.com "infinispan-operator.v0.1.0" is invalid: []: Invalid value: map[string]interface {}{"kind":"ClusterServiceVersion", "apiVersion":"operators.coreos.com/v1alpha1", "metadata":map[string]interface {}{"creationTimestamp":"2019-03-19T09:54:56Z", "annotations":map[string]interface {}{"containerImage":"docker.io/jboss/infinispan-operator:0.1.0", "capabilities":"Basic Install", "certified":"false", "description":"TBD", "support":"Infinispan", "tectonic-visibility":"ocs", "categories":"Database", "createdAt":"2019-02-28 14:18:00", "alm-examples":"[\n  {\n    \"apiVersion\": \"infinispan.org/v1\",\n    \"kind\": \"Infinispan\",\n    \"metadata\": {\n      \"name\": \"example-infinispan\"\n    },\n    \"spec\": {\n      \"size\": 3\n    }\n  },\n  {\n    \"apiVersion\": \"infinispan.org/v1\",\n    \"kind\": \"Infinispan\",\n    \"metadata\": {\n      \"name\": \"example-infinispan\"\n    },\n    \"config\": {\n      \"sourceType\": \"ConfigMap\",\n      \"sourceRef\": \"infinispan-app-configuration\",\n      \"name\": \"cloud-ephemeral.xml\"\n    },\n    \"spec\": {\n      \"size\": 3\n    }\n  }\n]\n"}, "generation":1, "uid":"0d4bd25d-4a2d-11e9-bdf0-025036b333c2", "name":"infinispan-operator.v0.1.0", "namespace":"openshift-operators"}, "spec":map[string]interface {}{"version":"0.1.0", "displayName":"infinispan", "provider":map[string]interface {}{"name":"Infinispan"}, "maturity":"stable", "labels":map[string]interface {}{"operated-by":"infinispan-operator"}, "apiservicedefinitions":map[string]interface {}{}, "description":"TBD\n\nTBD\n\nTBD\n", "installModes":[]interface {}{map[string]interface {}{"supported":true, "type":"OwnNamespace"}, map[string]interface {}{"supported":true, "type":"SingleNamespace"}, map[string]interface {}{"type":"MultiNamespace", "supported":true}, map[string]interface {}{"type":"AllNamespaces", "supported":true}}, "selector":map[string]interface {}{"matchLabels":map[string]interface {}{"operated-by":"infinispan-operator"}}, "install":map[string]interface {}{"strategy":"deployment", "spec":map[string]interface {}{"deployments":[]interface {}{map[string]interface {}{"name":"infinispan-operator", "spec":map[string]interface {}{"template":map[string]interface {}{"metadata":map[string]interface {}{"labels":map[string]interface {}{"name":"infinispan-operator-alm-owned"}, "name":"infinispan-operator-alm-owned"}, "spec":map[string]interface {}{"containers":[]interface {}{map[string]interface {}{"command":[]interface {}{"infinispan-operator"}, "env":[]interface {}{map[string]interface {}{"name":"WATCH_NAMESPACE", "valueFrom":map[string]interface {}{"fieldRef":map[string]interface {}{"fieldPath":"metadata.namespace"}}}, map[string]interface {}{"name":"POD_NAME", "valueFrom":map[string]interface {}{"fieldRef":map[string]interface {}{"fieldPath":"metadata.name"}}}, map[string]interface {}{"name":"OPERATOR_NAME", "value":"infinispan-operator"}}, "image":"jboss/infinispan-operator:0.1.0", "name":"infinispan-operator", "ports":[]interface {}{map[string]interface {}{"containerPort":60000, "name":"metrics"}}}}, "serviceAccountName":"infinispan-operator"}}, "replicas":1, "selector":map[string]interface {}{"matchLabels":map[string]interface {}{"name":"infinispan-operator-alm-owned"}}}}}, "permissions":[]interface {}{map[string]interface {}{"rules":[]interface {}{map[string]interface {}{"apiGroups":[]interface {}{""}, "resources":[]interface {}{"pods", "services", "endpoints", "persistentvolumeclaims", "events", "configmaps", "secrets"}, "verbs":[]interface {}{"*"}}, map[string]interface {}{"apiGroups":[]interface {}{""}, "resources":[]interface {}{"namespaces"}, "verbs":[]interface {}{"get"}}, map[string]interface {}{"apiGroups":[]interface {}{"apps"}, "resources":[]interface {}{"deployments", "daemonsets", "replicasets", "statefulsets"}, "verbs":[]interface {}{"*"}}, map[string]interface {}{"apiGroups":[]interface {}{"monitoring.coreos.com"}, "resources":[]interface {}{"servicemonitors"}, "verbs":[]interface {}{"get", "create"}}, map[string]interface {}{"apiGroups":[]interface {}{"infinispan.org"}, "resources":[]interface {}{"*"}, "verbs":[]interface {}{"*"}}}, "serviceAccountName":"infinispan-operator"}}}}, "customresourcedefinitions":map[string]interface {}{"owned":[]interface {}{map[string]interface {}{"name":"infinispans.infinispan.org", "version":"v1", "kind":"Infinispan", "displayName":"TBD", "description":"TBD.", "resources":[]interface {}{map[string]interface {}{"name":"", "kind":"Service", "version":"v1"}, map[string]interface {}{"name":"", "kind":"Pod", "version":"v1"}}, "specDescriptors":[]interface {}{map[string]interface {}{"path":"size", "displayName":"Size", "description":"TBD."}}}}}, "keywords":[]interface {}{"infinispan", "key value", "database", "datagrid", "open source"}, "maintainers":[]interface {}{map[string]interface {}{"name":"Infinispan Community", "email":"infinispan-dev@lists.jboss.org"}}, "links":[]interface {}{map[string]interface {}{"name":"Blog", "url":"https://blog.infinispan.org/"}, map[string]interface {}{"name":"Documentation", "url":"TBD"}, map[string]interface {}{"name":"etcd Operator Source Code", "url":"https://github.com/infinispan/infinispan-operator"}}}}: validation failure list:
        spec.links.url in body must be of type uri: "TBD"
```

The error is: `spec.links.url in body must be of type uri: "TBD"`

Got to this via `oc get installplan --all-namespaces -o yaml`

galderz commented:
After fixing that I've been able to install operator using UI.

Next I'm trying to create an example operator deployment by doing, but no pods are created:

```
$ oc project myproject
$ oc apply -f https://raw.githubusercontent.com/infinispan/infinispan-operator/0.1.0/deploy/cr/cr_minimal.yaml
infinispan.infinispan.org/example-infinispan created

$ oc get pods
No resources found.
```

I see the operator is up and running thought it had some restarts:

```
$ oc get pods -n openshift-operators
NAME                                   READY     STATUS    RESTARTS   AGE
infinispan-operator-56f8c8b669-svq4g   1/1       Running   3          26m
```

I've managed to get some logs when it's restarted:

```
{"level":"info","ts":1552993393.6632445,"logger":"cmd","msg":"Registering Components."}
{"level":"info","ts":1552993393.6635504,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"infinispan-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1552993393.6637378,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"infinispan-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1552993393.6638527,"logger":"cmd","msg":"Starting the Cmd."}
{"level":"info","ts":1552993393.764399,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"infinispan-controller"}
{"level":"info","ts":1552993393.8646677,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"infinispan-controller","worker count":1}
E0319 11:08:28.395709       1 streamwatcher.go:109] Unable to decode an event from the watch stream: http2: server sent GOAWAY and closed the connection; LastStreamID=137, ErrCode=NO_ERROR, debug=""
E0319 11:08:28.395709       1 streamwatcher.go:109] Unable to decode an event from the watch stream: http2: server sent GOAWAY and closed the connection; LastStreamID=137, ErrCode=NO_ERROR, debug=""
log: exiting because of error: log: cannot create log: open /tmp/infinispan-operator.infinispan-operator-56f8c8b669-svq4g.unknownuser.log.ERROR.20190319-110828.1: no such file or directory
```

Any idea about no pods appearing? And about the watcher error?

galderz commented:
Actually, it's all working fine. I seem to have created the CRD deployment too quickly after installing the operator and it didn't get through the worker process. After calling `oc apply ...` again it works fine :)

galderz commented:
Although this seems to work, I see the operator having many restarts (maybe due to issue above)? 

I've also seem something odd where the operator had been instantiated in one namespace (`myproject`), but somehow after today (not sure yesterday), the Infinispan instances appeared in the `openshift-operators` namespace.

galderz commented:
I'm now trying to test it on minikube

galderz commented:
I'm unable to install the marketplace operator on minikube (see https://github.com/operator-framework/operator-marketplace/issues/142 and https://github.com/operator-framework/operator-marketplace/issues/143).

Can someone help with this and the comments above?

We've added all the metadata for the time being.

galderz commented:
And can someone move to this to test to see if there's something we've missed>

SamiSousa commented:
@galderz Great work with debugging on your own! Sorry for getting back so late.
As for installing marketplace on minikube, if you use the `--validate=false` flag with kubectl, it should install regardless of the invalid field.
As for your question on OpenShift, the UI flow for installing includes a step where you select the operator's target namespace. By default if the `AllNamespaces` installMode is supported, the operator will be installed into `openshift-operators`, but if it supports the `SingleNamespace` installMode you can install it into another namespace.

galderz commented:
> As for installing marketplace on minikube, if you use the --validate=false flag with kubectl, it should install regardless of the invalid field.

Yeah, I wasn't so much worried about operator-marketplace#142, but operator-marketplace#143 was a showstopper. I'll try again.

> As for your question on OpenShift, the UI flow for installing includes a step where you select the operator's target namespace. By default if the AllNamespaces installMode is supported, the operator will be installed into openshift-operators, but if it supports the SingleNamespace installMode you can install it into another namespace.

In the metadata I submitted we've put:

```
  installModes:
    - type: OwnNamespace
      supported: true
    - type: SingleNamespace
      supported: true
    - type: MultiNamespace
      supported: true
    - type: AllNamespaces
      supported: true
```

When I try to install it that the UI it seems to say that it's only possible to install it in all namespaces (see [here](https://www.dropbox.com/s/hq5dfsulw4rmkyp/Screen%20Shot%202019-03-28%20at%2015.59.39.png?dl=0)). It says that a specific namespace is not supported by this operator in spite of the metadata above, why is that?

Related to this, even if the operator is installed in `openshift-operators` namespace meaning it's available to all, why is it that if I go to a different namespace, e.g. `myproject` and I try to instantiate the operator no pods are created there? If I look in `myproject` I do see the `infinispan` kind resource instance but not the pods. Only when I go to `openshift-operators` namespace and instantiate the operator there I see both the pods and the `infinispan` kind resource instance there. Is this a metadata problem or a problem with the operator code itself?

galderz commented:
@SamiSousa I've a couple of changes to:

- Upgrade to version 0.2.0 of the Infinispan operator. We have an issue with it and in the next few days we expect to release 0.2.1, which would be ready for operatorhub.
- Disabled `AllNamespaces`. After doing that it all seems to work fine. I can select the namespace on which to install it and works as expected.
- Fixed `WATCH_NAMESPACE` field value.
- Added Infinispan icon information.

Once again I've verified it all works fine on OpenShift 4. Next week I'll try with Minikube again.

galderz commented:
@SamiSousa I'm having some issues with testing on minikube. I've managed to install OLM and marketplace, and I've created all operatorsource, catalogsource and subscription (for local-operators namespace), but when if I go to the local-operators namespace and try to instantiate the operator I get:

```
$ kubectl apply -f https://raw.githubusercontent.com/infinispan/infinispan-operator/0.1.0/deploy/cr/cr_minimal.yaml -n local-operators
error: unable to recognize "https://raw.githubusercontent.com/infinispan/infinispan-operator/0.1.0/deploy/cr/cr_minimal.yaml": no matches for kind "Infinispan" in version "infinispan.org/v1"
```

The Infinispan CRD does not appear in the `kubectl get crd --all-namespaces` list. Am I missing something here? The [document you linked](https://github.com/operator-framework/operator-marketplace#deploying-the-marketplace-operator-with-kubernetes) stops short of what needs to be done after adding the Subscription.

galderz commented:
Based on what I say with OpenShift 4, I suspect an `InstallPlan` might be missing here. I don't see any for the Infinispan operator in my minikube instance.

galderz commented:
Found the mistake, I had created the subscription with a different channel (`alpha`) to the channel of the Infinispan operator (`stable`).

Even though I've now applied the example Infinispan yaml, the pods didn't appear. I can see in the event log:

```
local-operators   3m39s       Warning   NoOperatorGroup     ClusterServiceVersion   csv in namespace with no operatorgroups
```

Guess I'm missing the OperatorGroup, how do I add one?

SamiSousa commented:
@galderz Here is an OLM doc for OperatorGroups: https://github.com/operator-framework/operator-lifecycle-manager/blob/master/Documentation/design/operatorgroups.md#target-namespace-selection
In your case, you only want one targetNamespace which matches the namespace the operator is being installed into. 
If you used the OpenShift web console UI, the OperatorGroup should have been created for you, but I understand through debugging you may have moved to a more manual approach

galderz commented:
After looking around found [this issue](https://github.com/operator-framework/operator-sdk-samples/issues/57), which crucially led to [piece of code](https://github.com/estroz/community-operators/blob/scorecard-pr-check/scripts/lib/file#L50). Once I added the operator group it all worked fine on Minikube.

galderz commented:
Next I need to update our submission to operator version 0.2.1 since we o an important issue with 0.2.1. Once that's working fine, I will push it to an `infinispan` quay.io repo (currently use my own). Once that's all in place this PR won't be WIP any more.

galderz commented:
FYI, I've created a [PR](https://github.com/infinispan/infinispan-operator/pull/35/files) in our repo to test quay.io operatorhub.io submissions based on descriptors in this PR on minikube. It goes through the entire process from scratch `minikube` and then tests that a cluster of 3 Infinispan instances can be created using an Infinispan CR instantiation (having 3 instances is what makes minikube being configured with 4 CPUs and 4gb memory).

galderz commented:
@SamiSousa This is ready now to go in. I've tested it with Minikube and it runs fine. Can you please integrate it?

I have some questions now:

* Once integrated, would it be pushed to https://quay.io/application/upstream-community-operators/infinispan-operator repository? Or do we have to take care of this?
* Do I have to wait for this to be integrated to submit a community-operator version for this?
* Once we have Infinispan in community-operator, is there anything else we have to do so that it appears by default on the OpenShift 4 OperatorHub section?

galderz commented:
Actually, looking at the strimzi model, I think it'll be easier if the csv file and the name does not contain version information in it. I'll apply those changes and retest.

galderz commented:
Reverted back last commit, the version should be present in file.

Last CI run was failing with:

```
ERROR: the server was unable to return a response in the time allotted, but may still be processing the request
ERROR: Get https://10.30.6.20:8443/apis/infinispan.org/v1/namespaces/infinispan/infinispans/example-infinispan: dial tcp 10.30.6.20:8443: connect: connection refused
```

Any idea what would be wrong? Or is the error related to something other than this PR?

SamiSousa commented:
@AlexNPavel Here's the logs from scorecard: https://travis-ci.com/operator-framework/community-operators/builds/107191295#L1274
I'm not sure if this is the operator or a fluke, rerunning tests to see

galderz commented:
@SamiSousa I've noticed the scorecard messages, but I'm not sure how to fix them. Is there an operator that passes this scorecard? e.g. I don't see all the operators in this github repo with status in their CRDs...

galderz commented:
e.g.

```
SUGGESTION: Add a 'status' field to your Custom Resource
SUGGESTION: Add resources to owned CRDs
SUGGESTION: Add a spec descriptor for image
```

Any operator with examples of those?

SamiSousa commented:
@galderz The status section is added to CRs (instances of CRDs) once the object is created on cluster. It's a sign that the operator is watching the CRs when they are created.

galderz commented:
@SamiSousa Ok, thanks for clarifying that. We'll add it in forthcoming operator versions.

Is the `Add resources to owned CRDs` about the operator code itself too? Maybe some resources are not owned?

The last one is pretty straightforward.

galderz commented:
@SamiSousa I assume that the scorecard suggestions are not a blocker here to get this PR integrated? I mean, those suggestions don't make the CI build fail ;)

I'm tracking those separately and I'm working to fix them too.

Anything else missing here?

SamiSousa commented:
@galderz I am getting the following error logs from the operator pod when I create the example CR:
```
{"level":"info","ts":1554755982.8702178,"logger":"controller_infinispan","msg":"Creating a new Deployment","Request.Namespace":"default","Request.Name":"example-infinispan","Deployment.Namespace":"default","Deployment.Name":"example-infinispan"}
{"level":"info","ts":1554755982.9863763,"logger":"controller_infinispan","msg":"Reconciling Infinispan","Request.Namespace":"default","Request.Name":"example-infinispan"}
{"level":"error","ts":1554755983.0287411,"logger":"controller_infinispan","msg":"failed to update Deployment","Request.Namespace":"default","Request.Name":"example-infinispan","Deployment.Namespace":"default","Deployment.Name":"example-infinispan","error":"Operation cannot be fulfilled on statefulsets.apps \"example-infinispan\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/infinispan/infinispan-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/infinispan/infinispan-operator/pkg/controller/infinispan.(*ReconcileInfinispan).Reconcile\n\t/go/src/github.com/infinispan/infinispan-operator/pkg/controller/infinispan/infinispan_controller.go:148\ngithub.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:213\ngithub.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
{"level":"error","ts":1554755983.034326,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"infinispan-controller","request":"default/example-infinispan","error":"Operation cannot be fulfilled on statefulsets.apps \"example-infinispan\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/infinispan/infinispan-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
```
The operator doesn't crash though, and is able to create two pods for the example CR, with a 3rd pod stuck due to limited cpu on my cluster.

Also, what resources should the CR be creating? The CR doesn't seem to list these resources, so I'm having trouble checking that everything is set up properly

gfbarros commented:
Please add a repository to your csv. Here is the output of operator-courier:
operator-courier verify upstream-community-operators/infinispan/ --ui_validate_io
WARNING:operatorcourier.validate:csv metadata.annotations.repository not defined.Without this field, the link to the operator source code will not be displayed in the UI.

gfbarros commented:
Another quick note, when I ran the example cr w scale:2 i seem to have run out of resources on a 2gb,2core minikube setup.

galderz commented:
@SamiSousa I've been trying locally and can't see the error you mentioned above (`Operation cannot be fulfilled on statefulsets.apps`). I'm on minikube 1.0.0 (kubectl 1.14.0). What about you?

@gfbarros operator-courier didn't show up that `WARNING` when used without `--ui_validate_io`, why is not that default? I'll add that.

@gfbarros What resources are you running out of exactly? Do you have other stuff running on your minikube profile? I've been able to run fine the example with spec.size=2. I think 2 is a good minimum since you can then verify that a cluster forms between them (internal cluster). If you want I can switch the example in the CSV to spec.size=1, but I'd rather keep it in 2.

gfbarros commented:
> @gfbarros What resources are you running out of exactly? Do you have other stuff running on your minikube profile? I've been able to run fine the example with spec.size=2. I think 2 is a good minimum since you can then verify that a cluster forms between them (internal cluster). If you want I can switch the example in the CSV to spec.size=1, but I'd rather keep it in 2.

@galderz I'm guessing memory as it locks up the environment. Just using stock minikube setup with nothing else running.

galderz commented:
@gfbarros Rather than guess, maybe you can use `kubectl get events --all-namespaces` and `kubectl describe node minikube` to see exactly what is happening.

galderz commented:
Added a new commit to add spec image description, set example spec size to 2, and add repository information.

gfbarros commented:
@galderz I had no luck sorting out minikube so i moved to a larger OCP4 cluster and tried it there with no problems. I suspect my 2core/2gb minikube was just having trouble keeping up. 

The CR seems to apply correctly and here are the only warnings/errors I got. Please let me know if these are expected:
in both 'kubectl logs example-infinispan-0' & 'kubectl logs example-infinispan-1'
00:28:48,663 WARN  [org.jboss.as.txn] (ServerService Thread Pool -- 31) WFLYTX0013: The node-identifier attribute on the /subsystem=transactions is set to the default value. This is a danger for environments running multiple servers. Please make sure the attribute value is unique.
...
00:28:50,050 WARN  [org.jboss.as.domain.management.security] (MSC service thread 1-2) WFLYDM0111: Keystore /opt/jboss/infinispan-server/standalone/configuration/application.keystore not found, it will be auto generated on first use with a self signed certificate for host localhost
...
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.thoughtworks.xstream.core.util.Fields (jar:file:/opt/jboss/infinispan-server/modules/system/add-ons/ispn/com/thoughtworks/xstream/ispn-10.0/xstream-1.4.10.jar!/) to field java.util.TreeMap.comparator
WARNING: Please consider reporting this to the maintainers of com.thoughtworks.xstream.core.util.Fields
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

in 'kubectl logs infinispan-operator-7857c656-wq4jh'
{"level":"error","ts":1554942407.5870347,"logger":"controller_infinispan","msg":"failed to update Deployment","Request.Namespace":"default","Request.Name":"example-infinispan","Deployment.Namespace":"default","Deployment.Name":"example-infinispan","error":"Operation cannot be fulfilled on statefulsets.apps \"example-infinispan\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/infinispan/infinispan-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/infinispan/infinispan-operator/pkg/controller/infinispan.(*ReconcileInfinispan).Reconcile\n\t/go/src/github.com/infinispan/infinispan-operator/pkg/controller/infinispan/infinispan_controller.go:148\ngithub.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:213\ngithub.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
{"level":"error","ts":1554942407.5950313,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"infinispan-controller","request":"default/example-infinispan","error":"Operation cannot be fulfilled on statefulsets.apps \"example-infinispan\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/infinispan/infinispan-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/infinispan/infinispan-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}

galderz commented:
>  I had no luck sorting out minikube so i moved to a larger OCP4 cluster and tried it there with no problems. I suspect my 2core/2gb minikube was just having trouble keeping up.

@gfbarros I'd love to help since I don't see such issues locally, but as I said in my last comment, I need some info from you to be able to help further. My suspicion is that you have other stuff running there apart, but can't say unless you share some info :)

Those WARN messages look ok, they're part of the inner server logging. 

I see you are also see `Operation cannot be fulfilled on statefulsets.apps` errors, but I've been unable to replicate those locally on either a Minikube or an OpenShift 4 cluster. Even though they seem noisy, does it eventually get reconciled? IOW, I'd expect those errors will probably stop at some point. I've just looked into our CI logs ([here](https://travis-ci.org/infinispan/infinispan-operator/jobs/518389314)) and I see them appearing a couple of times, but doesn't seem to cause any real issues.

gfbarros commented:
> I see you are also see `Operation cannot be fulfilled on statefulsets.apps` errors, but I've been unable to replicate those locally on either a Minikube or an OpenShift 4 cluster. Even though they seem noisy, does it eventually get reconciled? IOW, I'd expect those errors will probably stop at some point. I've just looked into our CI logs ([here](https://travis-ci.org/infinispan/infinispan-operator/jobs/518389314)) and I see them appearing a couple of times, but doesn't seem to cause any real issues.

@galderz yes it looks like they stop and I then see Reconciling messages:
{"level":"info","ts":1554942408.5953696,"logger":"controller_infinispan","msg":"Reconciling Infinispan","Request.Namespace":"default","Request.Name":"example-infinispan"}
...
{"level":"info","ts":1554942538.6007683,"logger":"controller_infinispan","msg":"Reconciling Infinispan","Request.Namespace":"default","Request.Name":"example-infinispan"}

So it sounds to me that the messages are ok and everything is passing on OCP, do you agree?

gfbarros commented:
@galderz Do you have specific resource requirements for the app/operator? Could you list them in the CSV's spec.description section?

gfbarros commented:
@galderz one more question. It looks like you only have files in the upstream-community-operators folder. In order for it to show up on the OpenShift site you also need a copy of the same files in the community-operators folder. It's up to you which option you choose, just wanted to give you the option.

gfbarros commented:
@galderz one more question. It looks like you only have files in the upstream-community-operators folder. In order for it to show up on the OpenShift site you also need a copy of the same files in the community-operators folder. It's up to you which option you choose, just wanted to give you the option.

gfbarros commented:
@galderz one more question. It looks like you only have files in the upstream-community-operators folder. In order for it to show up on the OpenShift site you also need a copy of the same files in the community-operators folder. It's up to you which option you choose, just wanted to give you the option.

gfbarros commented:
@galderz one more question. It looks like you only have files in the upstream-community-operators folder. In order for it to show up on the OpenShift site you also need a copy of the same files in the community-operators folder. It's up to you which option you choose, just wanted to give you the option.

gfbarros commented:
@galderz one more question. It looks like you only have files in the upstream-community-operators folder. In order for it to show up on the OpenShift site you also need a copy of the same files in the community-operators folder. It's up to you which option you choose, just wanted to give you the option.

galderz commented:
@gfbarros Do you mean compute resources like we discussed above? Or other resources?

Yeah, we also want to appear in OpenShift site. I'll add files to community-operators too. Shall I use a different PR for that?

Btw, I have a couple of doubts remaining:

> `SUGGESTION: Add resources to owned CRDs` in the scorecard. 

What exactly is this expecting?

@SamiSousa's comment:

> "Also, what resources should the CR be creating? The CR doesn't seem to list these resources, so I'm having trouble checking that everything is set up properly"

What exactly is that?

Any examples on these would help!

gfbarros commented:
@galderz Go ahead and use the same PR to add the community-operators files.

Yes, compute resources. I'll get back to sorting out minikube and why it's acting up but that has caused some internal discussion on wanting operators that use "more" amounts of memory or cpu to mention in their spec.description what to expect.

galderz commented:
Added information to spec.description on resources. Also added community-operator files.

@gfbarros I'm still intrigued how you cannot run with default Minikube settings when pod number is 2. Could you share some info to see if we spot something?

galderz commented:
Also, I still have doubts what this is about:

> `SUGGESTION: Add resources to owned CRDs` in the scorecard.

AlexNPavel commented:
@galderz That test is overly strict at the moment. I will be changing how that works in a future PR, so that the test itself passes if the resources field exists and then mentions in the suggestions if it identified any other resources that you may want to add in the resources field. Currently it fails if any resources modified were omitted in the resources section, which is not how it should behave. Your CSV would be fine when the scorecard gets that fix in.

gfbarros commented:
@galderz here is how we are testing with OLM: https://github.com/operator-framework/community-operators/blob/master/docs/testing-operators.md#manual-testing-on-kubernetes

Could you run through the test using minikube and let me know if it works for you? And if so, maybe what we may be doing wrong? If you have tested with OLM as described above and are comfortable that it works please let us know as well.

SamiSousa commented:
@galderz Not sure why, but the scorecard doesn't seem to play nice with this operator :confused: 
When running the operator on my machine and adding the example CR, I see that some example_infinispan pods come up in response. However, it would be better for UX if the CR's status were updated to reflect that it was reconciled by the operator. Checking the operator's logs confirms this though.
@gfbarros It works for me, the operator pod logs look good:
```
{"level":"info","ts":1556580413.256922,"logger":"cmd","msg":"Registering Components."}
{"level":"info","ts":1556580413.281527,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"infinispan-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1556580413.2840939,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"infinispan-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1556580413.2850788,"logger":"cmd","msg":"Starting the Cmd."}
{"level":"info","ts":1556580413.893506,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"infinispan-controller"}
{"level":"info","ts":1556580413.9958668,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"infinispan-controller","worker count":1}
{"level":"info","ts":1556580413.9964414,"logger":"controller_infinispan","msg":"Reconciling Infinispan","Request.Namespace":"infinispan","Request.Name":"example-infinispan"}
```
/lgtm

gfbarros commented:
Yeah, I'm happy if you and @galderz are happy. It works great on OpenShift for me but just borks my minikube setup. 

/lgtm

galderz commented:
@SamiSousa @gfbarros Thanks for integrating this.
